### PR TITLE
Created attributes.js

### PR DESCRIPTION
**Pull request by bradfrizzell**
This new file provides the getContextMetadata and getNameAttribute functions. getContextMetadata is a generalized function for assigning the metadata needed for setting up window.context

This addition was part of https://github.com/ampproject/amphtml/pull/6825 but I've realized that PR had a mixture of things that could be better suited to review by breaking apart further.

### CODE REVIEW

**Comment by jridgewell** [comment]
These are both 3p iframe specific. Maybe put in the 3p directory?

**Comment by bradfrizzell** [comment]
These will also be used in the A4A flow for non-amp ads, so it is not only for 3p frames.

**Comment by jridgewell** [comment]
> These will also be used in the A4A flow for non-amp ads, so it is not only for 3p frames.
View changes

But that just creates a 3p frame... right?

**Comment by bradfrizzell** [comment]
No, although it is rendering a non-amp ad, it doesn't actually use a 3p-frame. https://github.com/ampproject/amphtml/blob/master/extensions/amp-a4a/0.1/amp-a4a.js#L1077

File: src/attributes.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Produces the attributes for the ad template.
+ * @param {!Window} parentWindow
+ * @param {!Element} element
+ * @return {!Object} Contains
+ *       precedence over the data- attributes.
+ *     - data-* attributes of the <amp-ad> tag with the "data-" removed.
+ *     - A _context object for internal use.
+ */
+export function getContextMetadata(
```
Comment (jridgewell) [inline src/attributes.js]: What code can be deleted from https://github.com/ampproject/amphtml/blob/master/src/3p-frame.js?

Comment (bradfrizzell) [inline src/attributes.js]: I can remove a fair amount (which I have done, but it's not in this pr). If you'd like that all to be in this PR, I can do that, but getting this attributes.js file in for the next release is important for us fixing an issue with a4a so I didn't want to slow down if there were any issues integrating this file into 3p-frame.js

Comment (jridgewell) [inline src/attributes.js]: As long as there's a quick follow up. I hate adding duplicated code.

> for the next release is important for us fixing an issue with a4a

Where's the code using this?

Comment (bradfrizzell) [inline src/attributes.js]: https://github.com/ampproject/amphtml/pull/7132

**Comment by bradfrizzell** [comment]
this blocks https://github.com/ampproject/amphtml/pull/7132

**Comment by jridgewell** [comment]
LGTM, @lannka?

**Comment by bradfrizzell** [comment]
Would like to get this into today's release

File: src/iframe-attributes.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    startTime,
+  };
+  attributes._context[sentinelNameChange ? 'sentinel' : 'amp3pSentinel'] =
+      sentinel;
+  const adSrc = element.getAttribute('src');
+  if (adSrc) {
+    attributes.src = adSrc;
+  }
+  return attributes;
+
```
Comment (lannka) [inline src/iframe-attributes.js]: nit: remove blank line

**Comment by bradfrizzell** [comment]
I want it in this release so that I can also get #7132 in today's release (which this is blocking)

**Comment by bradfrizzell** [comment]
Refactored it as requested

**Comment by lannka** [comment]
Thanks @bradfrizzell . were there any tests need to be moved as well?

**Comment by bradfrizzell** [comment]
I think the tests in test-3p-frame provide good testing of the getContextMetadata, there was nothing in the other PR already written that wasn't moved at least. I can add more if you think getContextMetadata isn't well tested enough. In the pr that uses it for a4a, there are tests of it's use for the a4a case that will cover it more

**Comment by lannka** [comment]
@bradfrizzell I think it make a lot sense to move those tests from `test-3p-frame.js` to `test-attributes.js`, since the major logic is moved.

**Comment by bradfrizzell** [comment]
reverted the last two commits that I made to break out the tests, because I was getting really weird errors with state leakage in the tests. basically, test-iframe-attributes.js was passing if I ran it alone, but if I ran it as part of the suite as a whole it was not. Getting this PR into today's cut is of very high importance to the A4A team, so I need to push off breaking this test apart if we can

**Pull request merged**